src/clean-core/string_stream.hh

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <clean-core/string.hh>
+#include <clean-core/string_view.hh>
+#include <cstring> // std::memcpy
+
+namespace cc
+{
+class string_stream
+{
+public: // methods
+    string_stream& operator<<(string_view sv)
+    {
+        reserve(sv.size());
+        std::memcpy(m_curr, sv.data(), sv.size());
+        m_curr += sv.size();
+        return *this;
+    }
+
+    [[nodiscard]] string to_string() const
+    {
+        string s = string::uninitialized(size());
+        memcpy(s.data(), m_data, size());
+        return s;
+    }
+
+    /// reseve space for at least size more elements
+    void reserve(size_t size)
+    {
+        size_t req_size = this->size() + size;
+        if (req_size <= m_capacity)
+            return;
+
+        size_t new_cap = (m_capacity << 1) < req_size ? req_size : (m_capacity << 1);
+        char* new_data = new char[new_cap];
+        std::memcpy(new_data, m_data, this->size());
+        delete[] m_data;
+        m_curr = m_curr - m_data + new_data;
+        m_data = new_data;
+        m_capacity = new_cap;
+    }
+
+    void clear() { m_curr = m_data; }
+
+public: // properties
+    [[nodiscard]] size_t size() const noexcept { return m_curr - m_data; }
+    [[nodiscard]] bool empty() const { return m_curr == m_data; }
+
+public: // ctor
+    string_stream() = default;
+
+    string_stream(string_stream const& rhs)
+    {
+        m_data = new char[rhs.size()];
+        std::memcpy(m_data, rhs.m_data, rhs.size());
+        m_curr = m_data + rhs.size();
+        m_capacity = rhs.size();
+    };
+
+    string_stream(string_stream&& rhs) noexcept
+    {
+        m_data = rhs.m_data;
+        m_curr = rhs.m_curr;
+        m_capacity = rhs.m_capacity;
+        rhs.m_data = nullptr;
+        rhs.m_curr = nullptr;
+        rhs.m_capacity = 0;
+    };
+
+    ~string_stream() { delete[] m_data; }
+
+public: // assignment
+    string_stream& operator=(string_stream const& rhs)
+    {
+        if (this != &rhs)
+        {
+            if (m_capacity < rhs.size())
+            {
+                delete[] m_data;
+                m_data = new char[rhs.size()];
```

Comment:
this behaves unintuitive if `rhs` is empty (size will be 0 but m_data won't be `nullptr`)